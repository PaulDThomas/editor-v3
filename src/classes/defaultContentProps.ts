import { EditorV3Align, EditorV3ContentProps } from "./interface";
import { defaultMarkdownSettings } from "./markdown/MarkdownSettings";

export const defaultContentProps: EditorV3ContentProps = {
  allowMarkdown: false,
  allowNewLine: false,
  decimalAlignPercent: 60,
  markdownSettings: defaultMarkdownSettings,
  showMarkdown: false,
  styles: undefined,
  textAlignment: EditorV3Align.left,
  atListFunction: undefined,
  maxAtListLength: 10,
};
