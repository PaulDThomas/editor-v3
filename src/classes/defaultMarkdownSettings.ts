export interface IMarkdownSettings {
  styleStartTag: string;
  styleNameEndTag: string;
  styleEndTag: string;
  defaultStyle: string;
  dropDownStartTag: string;
  dropDownEndTag: string;
  dropDownValueSeparator: string;
  dropDownSelectedValueTag: string;
  calcStartTag: string;
  calcNameEndTag: string;
  calcEndTag: string;
  calcConditionSeparator: string;
  calcAndSeparator: string;
  atStartTag: string;
  atEndTag: string;
}

export const defaultMarkdownSettings = {
  styleStartTag: "<<",
  styleNameEndTag: "::",
  styleEndTag: ">>",
  defaultStyle: "defaultStyle",
  dropDownStartTag: "[[",
  dropDownEndTag: "]]",
  dropDownValueSeparator: "||",
  dropDownSelectedValueTag: "**",
  calcStartTag: "[![",
  calcNameEndTag: "::",
  calcEndTag: "]!]",
  calcConditionSeparator: "||",
  calcAndSeparator: "&&",
  atStartTag: "@[",
  atEndTag: "@]",
};
