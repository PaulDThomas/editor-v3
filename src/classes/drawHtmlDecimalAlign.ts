import { EditorV3BlockClass, EditorV3ContentProps, EditorV3RenderProps } from "./interface";

export const drawHtmlDecimalAlign = (
  renderProps: EditorV3RenderProps,
  contentProps: EditorV3ContentProps,
  lhsContent: EditorV3BlockClass[],
  rhsContent: EditorV3BlockClass[],
) => {
  if (renderProps.currentEl) {
    const prePoint = document.createElement("span");
    prePoint.className = "aiev3-span-point lhs";
    renderProps.currentEl.append(prePoint);
    if (lhsContent.length)
      lhsContent.forEach((tb) =>
        prePoint.append(
          tb.toHtml(
            { ...renderProps, currentEl: prePoint },
            tb.style ? contentProps.styles?.[tb.style] : undefined,
          ),
        ),
      );
    else prePoint.textContent = "\u200b";

    // Set up space after (excluding) decimal
    const postPoint = document.createElement("span");
    postPoint.className = "aiev3-span-point rhs";
    renderProps.currentEl.append(postPoint);
    if (rhsContent.length)
      rhsContent.forEach((tb) => {
        postPoint.append(
          tb.toHtml(
            { ...renderProps, currentEl: postPoint },
            tb.style ? contentProps.styles?.[tb.style] : undefined,
          ),
        );
      });
    else postPoint.textContent = "\u200b";

    // Set grid column widths
    renderProps.currentEl.style.gridTemplateColumns = `${contentProps.decimalAlignPercent}% ${100 - contentProps.decimalAlignPercent}%`;
  }
};
