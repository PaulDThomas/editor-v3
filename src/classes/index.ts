import { EditorV3Content } from "./EditorV3Content";
import { EditorV3Line } from "./EditorV3Line";
import { EditorV3AtBlock } from "./EditorV3AtBlock";
import { EditorV3TextBlock } from "./EditorV3TextBlock";
import { EditorV3Align, EditorV3Styles, IEditorV3, EditorV3DropListItem } from "./interface";

export { EditorV3Align, EditorV3Content, EditorV3Line, EditorV3TextBlock, EditorV3AtBlock };
export type { EditorV3Styles, IEditorV3, EditorV3DropListItem as EditorV3AtListItem };
