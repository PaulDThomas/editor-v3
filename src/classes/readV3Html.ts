import { EditorV3Line } from "./EditorV3Line";
import { EditorV3Align, EditorV3Import } from "./interface";
import { readV3DivElement } from "./readV3DivElement";

export const readV3Html = (
  text: string,
  textAlignment?: EditorV3Align,
  decimalAlignPercent?: number,
): EditorV3Import => {
  const ret: EditorV3Import = {
    lines: [],
  };

  // Read in v3-html string input
  if (text.match(/^<div class="aiev3-line.*">.*<\/div>$/)) {
    const frag = document.createElement("div");
    frag.innerHTML = text;

    // Read line nodes
    [...frag.querySelectorAll("div.aiev3-line")].forEach((el) => {
      const div = el as HTMLDivElement;
      ret.lines.push(readV3DivElement(div));
      // Get alignment from second classname
      if (!ret.textAlignment && div.classList.length > 1) {
        ret.textAlignment = el.classList[1].toString() as EditorV3Align;
      }
      // Get decimal align percent from mid width
      if (ret.textAlignment === EditorV3Align.decimal && el.children.length === 2) {
        ret.decimalAlignPercent = parseFloat((el.children[0] as HTMLSpanElement).style.minWidth);
      }
    });

    // Read style node
    const styleNode = frag.querySelector("div.aiev3-style-info");
    if (styleNode && (styleNode as HTMLDivElement).dataset.style !== "") {
      ret.styles = JSON.parse((styleNode as HTMLDivElement).dataset.style ?? "");
    }
  }
  // Everything else is just text(ish)
  else {
    ret.lines = text
      .replace(/[\u200B-\u200F\uFEFF\r\t]/g, "") // Undesirable non-printing chars
      .replace(/[\u202F|\u00A0]/g, " ") // Spaces are spaces
      .replaceAll("</div><div", "</div>\n<div")
      .replaceAll("&nbsp;", " ")
      .split("\n")
      .map((l) => new EditorV3Line(l, textAlignment, decimalAlignPercent));
  }
  return ret;
};
