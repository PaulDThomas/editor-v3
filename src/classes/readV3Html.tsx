import { EditorV3Line } from './EditorV3Line';
import { EditorV3Align, EditorV3Import } from './interface';

export const readV3Html = (
  text: string,
  textAlignment: EditorV3Align,
  decimalAlignPercent: number,
): EditorV3Import => {
  const ret: EditorV3Import = {
    lines: [],
  };

  // Read in v3-html string input
  if (text.match(/^<div class="aiev3-line.*">.*<\/div>$/)) {
    const frag = document.createElement('div');
    frag.innerHTML = text;

    // Loop through children in fragment
    [...frag.childNodes].forEach((el) => {
      ret.lines.push(new EditorV3Line(el.textContent ?? '', textAlignment, decimalAlignPercent));
    });
  }
  // Everything else is just text
  else {
    ret.lines = text
      .replace(/[\u200B-\u200F\uFEFF\r\t]/g, '') // Undesirable non-printing chars
      .replace(/[\u202F|\u00A0]/g, ' ') // Spaces are spaces
      .split('\n')
      .map((l) => new EditorV3Line(l, textAlignment, decimalAlignPercent));
  }
  return ret;
};
