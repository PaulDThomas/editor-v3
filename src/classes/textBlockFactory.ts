import {
  EditorV3AtBlock,
  IEditorV3AtBlock,
  IEditorV3AtBlockOptionalParams,
} from "./EditorV3AtBlock";
import {
  EditorV3TextBlock,
  IEditorV3TextBlock,
  IEditorV3TextBlockOptionalParams,
} from "./EditorV3TextBlock";

/**
 *  Factory function to direct text block creation
 * @param arg Block data/element
 * @param forcedParams Parameters to force
 * @returns EditorV3TextBlock | EditorV3AtBlock
 */
export const textBlockFactory = (
  arg:
    | HTMLSpanElement
    | Text
    | EditorV3TextBlock
    | DocumentFragment
    | IEditorV3TextBlock
    | IEditorV3AtBlock
    | string,
  forcedParams?: IEditorV3TextBlockOptionalParams | IEditorV3AtBlockOptionalParams,
): EditorV3TextBlock | EditorV3AtBlock => {
  // Span element
  if (arg instanceof HTMLSpanElement) {
    return arg.dataset.type === "at" || forcedParams?.type === "at"
      ? new EditorV3AtBlock(arg, forcedParams)
      : new EditorV3TextBlock(arg, forcedParams);
  }

  // Document Fragment element
  else if (arg instanceof DocumentFragment) {
    return arg.childNodes.length > 0 &&
      arg.childNodes[0] instanceof HTMLSpanElement &&
      (arg.childNodes[0].dataset.type === "at" || forcedParams?.type === "at")
      ? new EditorV3AtBlock(arg, forcedParams)
      : new EditorV3TextBlock(arg, forcedParams);
  }
  // Text node
  else if (arg instanceof Text) {
    return new EditorV3TextBlock(
      { text: (arg.textContent ?? "").replaceAll("\u00A0", " ") },
      forcedParams,
    );
  }
  // Text
  else if (typeof arg === "string") {
    try {
      const jsonInput = JSON.parse(arg);
      return jsonInput.type === "at" || forcedParams?.type === "at"
        ? new EditorV3AtBlock(jsonInput, forcedParams)
        : typeof jsonInput !== "object"
          ? new EditorV3TextBlock({ text: `${jsonInput}` }, forcedParams)
          : new EditorV3TextBlock(jsonInput, forcedParams);
    } catch {
      return new EditorV3TextBlock({ text: arg }, forcedParams);
    }
  }
  // Must be object
  else {
    return arg.type === "at" || forcedParams?.type === "at"
      ? new EditorV3AtBlock(arg, forcedParams)
      : new EditorV3TextBlock(arg, forcedParams);
  }
};
