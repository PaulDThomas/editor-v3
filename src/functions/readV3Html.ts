import { defaultContentProps } from "../classes/defaultContentProps";
import { EditorV3Line } from "../classes/EditorV3Line";
import { EditorV3ContentPropsInput, EditorV3Import } from "../classes/interface";
import { readContentPropsNode } from "./readContentPropsNode";
import { readV3DivElement } from "./readV3DivElement";
import { readV3MarkdownElement } from "./readV3MarkdownElement";

export const readV3Html = (
  text: string,
  contentProps?: EditorV3ContentPropsInput,
): EditorV3Import => {
  const ret: EditorV3Import = {
    lines: [],
    contentProps,
  };

  // Read in v3-html string input
  if (text.match(/^<div class="aiev3-line.*">.*<\/div>$/)) {
    const frag = document.createElement("div");
    frag.innerHTML = text;

    // Read content props node
    ret.contentProps = readContentPropsNode(frag);

    // Read line nodes
    [...frag.querySelectorAll("div.aiev3-line")].forEach((el) => {
      const div = el as HTMLDivElement;
      ret.lines.push(readV3DivElement(div));
    });
  }
  // Read in v3-markdown string input
  else if (text.match(/^<div class="aiev3-markdown-line.*">.*<\/div>$/)) {
    const frag = document.createElement("div");
    frag.innerHTML = text;

    // Read content props node
    ret.contentProps = readContentPropsNode(frag);

    // Read line nodes
    [...frag.querySelectorAll("div.aiev3-markdown-line")].forEach((el) => {
      const div = el as HTMLDivElement;
      const lineRet = readV3MarkdownElement(div, {
        ...defaultContentProps,
        ...ret.contentProps,
        ...contentProps,
      });
      ret.lines.push(lineRet);
    });
  }
  // Everything else is just text(ish)
  else {
    ret.lines = text
      .replace(/[\u2009-\u200F\r\t]/g, "") // Undesirable non-printing chars
      .replaceAll("</div><div", "</div>\n<div")
      .split("\n")
      .map((l) => new EditorV3Line(l, contentProps));
  }

  // Always take contentProps from the input if it exists
  if (contentProps) ret.contentProps = contentProps;
  return ret;
};
