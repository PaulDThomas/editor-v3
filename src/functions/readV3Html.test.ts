/* eslint-disable quotes */
import { defaultContentProps } from "classes/defaultContentProps";
import { textBlockFactory } from "../classes/textBlockFactory";
import { readV3Html } from "./readV3Html";

describe("readV3Html tests", () => {
  test("Read v3-html string input with line nodes", async () => {
    const text = `
      <div class="aiev3-line left"><span class="aiev3-tb">Hello world</span></div>
      <div class="aiev3-line left"><span class="aiev3-tb">How are you?</span></div>
      `
      .replaceAll(/[\r\n\t]/g, "")
      .replaceAll(/>\s{2,}</g, "><")
      .trim();
    const result = readV3Html(text, defaultContentProps);
    expect(result.lines.length).toEqual(2);
    expect(result.lines[0].textBlocks[0]).toEqual(textBlockFactory({ text: "Hello world" }));
    expect(result.lines[1].textBlocks[0]).toEqual(textBlockFactory({ text: "How are you?" }));
  });

  test("Read v3-markdown string input with line nodes", async () => {
    const text = `
      <div class="aiev3-markdown-line">(~(Hello world)~)</div>
      <div class="aiev3-markdown-line">(~(st1::How are you?)~)</div>
      `
      .replaceAll(/[\r\n\t]/g, "")
      .replaceAll(/>\s{2,}</g, "><")
      .trim();
    const result = readV3Html(text, defaultContentProps);
    expect(result.lines.length).toEqual(2);
    expect(result.lines[0].textBlocks).toEqual([
      textBlockFactory({ text: "Hello world", style: "defaultStyle" }),
    ]);
    expect(result.lines[1].textBlocks).toEqual([
      textBlockFactory({ text: "How are you?", style: "st1" }),
    ]);
  });

  test("Read text(ish) input", async () => {
    const text = "Hello world\nHow are you?";
    const result = readV3Html(text, defaultContentProps);
    expect(result.lines.length).toEqual(2);
    expect(result.lines[0].textBlocks).toEqual([textBlockFactory({ text: "Hello world" })]);
    expect(result.lines[1].textBlocks).toEqual([textBlockFactory({ text: "How are you?" })]);
  });

  test("Read in at block", async () => {
    const text = `
      <div class="aiev3-line left">
        <span class="aiev3-tb at-block" data-type="at" title="Hello@world">@Hello world</span>
      </div>
      `
      .replaceAll(/[\r\n\t]/g, "")
      .replaceAll(/>\s{1,}</g, "><")
      .trim();
    const result = readV3Html(text, defaultContentProps);
    expect(result.lines.length).toEqual(1);
    expect(result.lines[0].textBlocks).toMatchSnapshot();
  });
});
