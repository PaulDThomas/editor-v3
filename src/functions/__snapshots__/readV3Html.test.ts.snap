// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readV3Html tests Read in at block 1`] = `
[
  EditorV3AtBlock {
    "_defaultContentProps": {
      "allowMarkdown": false,
      "allowNewLine": false,
      "atListFunction": undefined,
      "decimalAlignPercent": 60,
      "markdownSettings": {
        "atEndTag": "@]",
        "atStartTag": "@[",
        "calcAndSeparator": "&&",
        "calcConditionSeparator": "||",
        "calcEndTag": "]!]",
        "calcNameEndTag": "::",
        "calcStartTag": "[![",
        "defaultStyle": "defaultStyle",
        "dropDownEndTag": "]]",
        "dropDownNameEndTag": "::",
        "dropDownSelectedValueTag": "**",
        "dropDownStartTag": "[[",
        "dropDownValueSeparator": "||",
        "styleEndTag": ">>",
        "styleNameEndTag": "::",
        "styleStartTag": "<<",
      },
      "showMarkdown": false,
      "styles": undefined,
      "textAlignment": "left",
    },
    "atData": {},
    "atListFunction": [Function],
    "isActive": false,
    "isLocked": undefined,
    "lineStartPosition": 0,
    "maxAtListLength": 10,
    "style": undefined,
    "text": "@Hello world",
    "type": "at",
  },
]
`;
